### PR TITLE
feat: update keycloak to use quarkus keycloak version 21-1-2

### PR DESCRIPTION
**Pull request by Zbozi**
# Ticket 🎫

This closes
- https://github.com/dot-base/keycloak/issues/64
- https://github.com/dot-base/keycloak/issues/65
- https://github.com/dot-base/keycloak/issues/69

Related PRs:
Changes in keycloak deployment: https://github.com/dot-base/deployments/pull/121

Keycloak config needs changes on all instances before releasing dotbase-23-4: https://github.com/dot-base/keycloak/issues/70

# Description 📖

* updates Keycloak to version  21.1 with quay.io base image
* Realm import on start only if no realm is existing
* Custom TOTP endpoints

# How to Test? 🧪

A test image based on the changes of this PR was built: [ghcr.io/dot-base/keycloak:PR-66](ghcr.io/dot-base/keycloak:PR-66
)

This image is deployed on [staging](staging.dotbase.charite.de) with the current dotbase 23-3 stack. See also: https://github.com/dot-base/deployments/pull/121

Steps to test manually can be found in #65 .

# Checklist ✅
- [x] My changes generate no new warnings
- [x] Made corresponding changes to the documentation if applicable
- [x] Commented my code, particularly in hard-to-understand areas if applicable
- [x] Performed a self-review of my own changes
- [ ] Added relevant unit tests

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -21,16 +21,10 @@ RUN cat /tmp/trust-ca/* >> /etc/ssl/certs/ca-bundle.crt
 RUN printf "\n\n# Manually added certs by dotbase\n" >> /etc/ssl/certs/ca-bundle.trust.crt
 RUN cat /tmp/trust-ca/* >> /etc/ssl/certs/ca-bundle.trust.crt
 
-USER jboss
+ENV KC_HTTPS_TRUST_STORE_FILE=/tmp/trust-ca/trust.jks
+ENV KC_HTTPS_TRUST_STORE_PASSWORD=password
```
Comment (cheyer) [inline Dockerfile:25]: Can we use a env variable for the password?

File: bin/kc_dotbase.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+source /opt/keycloak/kc.sh "start" "--import-realm" "--override false"
```
Comment (cheyer) [inline bin/kc_dotbase.sh:1]: We don't use it, it can be removed.

File: configuration/kc_dotbase_startup.sh [hunk 3]
```diff
@@ -0,0 +1 @@
+source /opt/keycloak/kc.sh "start" "--import-realm" "ignore-existing"
```
Comment (cheyer) [inline configuration/kc_dotbase_startup.sh:1]: We don't need it.

**Comment by cheyer** [comment]
When trying to get a QR for the patient app login, the QR code does not show. Also following error message pops up:
```
PATIENT_REALM_DATA_GRAPH_CLIENT_SECRET is undefined but required for the service If the error persists please contact our support team at movebase@charite.de.
```

File: Dockerfile (hunk 1 above)
Comment (Zbozi) [reply Dockerfile:25]: @andreakreichgauer wie stehst du dazu? In der alten Version war es auch hardcoded

File: Dockerfile [hunk 4]
```diff
@@ -4,15 +4,15 @@ COPY src/ /tmp/src/
...
+FROM quay.io/keycloak/keycloak:21.1.2 as builder
+RUN /opt/keycloak/bin/kc.sh build
 
-FROM jboss/keycloak:15.0.2
-COPY --from=build-totp  /tmp/target/dotbase_totp_provder-1.0-SNAPSHOT.jar /opt/jboss/keycloak/standalone/deployments/dotbase_totp_provder-1.0-SNAPSHOT.jar
-COPY configuration/standalone.xml /opt/jboss/keycloak/standalone/configuration/standalone.xml
-COPY realm-templates /opt/jboss/startup-scripts/
+FROM quay.io/keycloak/keycloak:21.1.2
+COPY --from=builder /opt/keycloak/ /opt/keycloak/
+COPY --from=build-totp /tmp/target/dotbase_totp_provder-1.0-SNAPSHOT.jar /opt/keycloak/providers/dotbase_totp_provder-1.0-SNAPSHOT.jar
```
Comment (Zbozi) [inline Dockerfile:12]: typo: i missing in `...totop_provder-1.0...`

File: src/main/java/org/dotbase/providers/TOTPProvider.java [hunk 5]
```diff
@@ -101,17 +100,12 @@ private UserModel isValidUser(RealmModel realm, String username) {
...
   }
 
-  private UserCredentialStore getCredentialStore() {
-    return session.userCredentialManager();
-  }
-
   public boolean configuredOTPCredential(RealmModel realm, UserModel user) {
-    if (getCredentialStore().getStoredCredentialsByTypeStream(realm, user, "totp").count() > 0) return true;
-    return getCredentialStore().getStoredCredentialsByTypeStream(realm, user, "otp").count() > 0;
+    return user.credentialManager().getStoredCredentialsByTypeStream("totp").count() > 0;
```
Comment (Zbozi) [inline src/main/java/org/dotbase/providers/TOTPProvider.java:104]:
```suggestion
   if(user.credentialManager().getStoredCredentialsByTypeStream("totp").count() > 0) return true;
   return user.credentialManager().getStoredCredentialsByTypeStream("otp").count() > 0;
```

Ich habe bei der Implementierung glaube ich was übersehen.

File: src/main/java/org/dotbase/providers/TOTPProvider.java [hunk 6]
```diff
@@ -101,17 +100,12 @@ private UserModel isValidUser(RealmModel realm, String username) {
     return user;
   }
 
-  private UserCredentialStore getCredentialStore() {
-    return session.userCredentialManager();
-  }
-
   public boolean configuredOTPCredential(RealmModel realm, UserModel user) {
```
[the remark this answers is not in the document]
Comment (andreakreichgauer) [reply src/main/java/org/dotbase/providers/TOTPProvider.java:102]: :+1:

File: .github/workflows/test-pr.yml
Comment (Zbozi) [inline .github/workflows/test-pr.yml:1]: Can be deleted?

File: .github/workflows/create-test-image.yml
Comment (Zbozi) [inline .github/workflows/create-test-image.yml:1]: Can be deleted?

**Pull request merged**